### PR TITLE
pr

### PR DESCRIPTION
**Pull request by L-as**
- [x] anabatic
 - [ ] bootstrap
 - [x] bora
 - [x] coloquinte
 - [x] crlcore
 - [x] cumulus
 - [ ] documentation
 - [x] equinox
 - [x] etesian
 - [x] flute
 - [x] hurricane
 - [x] ispd
 - [x] karakaze
 - [x] katabatic
 - [x] katana
 - [x] kite
 - [x] knik
 - [x] lefdef
 - [x] mauka
 - [x] metis
 - [x] nimbus
 - [x] oroshi
 - [x] solstice
 - [x] stratus1
 - [x] tutorial
 - [x] unicorn
 - [x] unittests
 - [x] vlsisapd

### CODE REVIEW

File: lefdef/src/lef/lef/CMakeLists.txt [hunk 1]
```diff
@@ -61,8 +61,8 @@
           add_custom_command ( OUTPUT   lef.tab.cpp
                                DEPENDS  ${LefParserGrammar}
                                COMMAND  ${BISON_EXECUTABLE} -v -p lefyy -d ${LefParserGrammar} -o lef.tab.cpp 
-                               COMMAND  mv lef.tab.hpp ${LEFDEF_SOURCE_DIR}/src/lef/lef/lef.tab.h
-                              #COMMAND  mv lef.tab.cpp ${LEFDEF_SOURCE_DIR}/src/lef/lef/lef.tab.cpp
+                               COMMAND  cp lef.tab.hpp ${LEFDEF_SOURCE_DIR}/src/lef/lef/lef.tab.h
+                               #COMMAND  mv lef.tab.cpp ${LEFDEF_SOURCE_DIR}/src/lef/lef/lef.tab.cpp
```
Comment (MatthewCroughan) [inline lefdef/src/lef/lef/CMakeLists.txt]: How come you are making changes to the `lefdef/src/lef/lef/CMakeLists.txt` file? Was this just for testing? Or do you intend to upstream this change in addition to the flake? In my mind, the correct way to do this is a patch.

Comment (L-as) [reply lefdef/src/lef/lef/CMakeLists.txt]: I'm not sure what I'm going to do with this flake: https://github.com/ngi-nix/summer-of-nix/discussions/44

In any case, I can convert the commit to a diff later (though the change would be useful upstream anyway).

**Comment by L-as** [comment]
I've made an overview of the components that need to be packaged.

File: nix/coloquinte.nix [hunk 2]
```diff
@@ -0,0 +1,18 @@
+{ version, meta }:
+
+{ lib, stdenv, python2, cmake, boost, qt4, ninja, coriolis-bootstrap }:
+
+let boostWithPython = boost.override { enablePython = true; python = python2; }; in
```
Comment (DieracDelta) [inline nix/coloquinte.nix]: They depend on python 2? That's mildly concerning. Saw the readme...

Comment (L-as) [reply nix/coloquinte.nix]: Libre-SoC seems to maybe use Python 3..? https://libre-soc.org/HDL_workflow/coriolis2/

I guess we'll find out when we try actually using it.

**Comment by L-as** [comment]
It's all packaged now, all that's left is testing it now.

**Comment by L-as** [comment]
It doesn't run, but I'm not going to debug this now.

File: nix/mauka.nix [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  buildInputs = [
+    python2 coriolis-bootstrap coriolis-vlsisapd coriolis-nimbus
+    coriolis-crlcore coriolis-hurricane qt4 boostWithPython
+    coriolis-metis
+  ];
+  nativeBuildInputs = [ cmake ninja doxygen ];
+
+  inherit version;
+
+  meta = meta // { broken = true; };
```
Comment (DieracDelta) [inline nix/mauka.nix]: is this broken?

Comment (L-as) [reply nix/mauka.nix]: Yep, it seems to be deprecated. It needs symbols from some other library in Coriolis (Hurricane IIRC) that aren't defined in the first place, so I assume it's neglected.

Comment (DieracDelta) [reply nix/mauka.nix]: ahh gotcha, thanks

**Comment by L-as** [comment]
I think it might work partially now. The `cgt` binary likely still doesn't run (haven't tested it), but I can import the Python extension modules from `unicorn`, etc.

**Comment by L-as** [comment]
https://github.com/NixOS/nixpkgs/pull/123263

**Comment by L-as** [comment]
Trying to get tests at https://gitlab.lip6.fr/vlsi-eda/alliance-check-toolkit working.

**Comment by L-as** [comment]
Fixed Alliance support in Nixpkgs: https://github.com/L-as/nixpkgs/commits/alliance

Perhaps @DieracDelta wants to take a look at it before I present it upstream? (I will also have to test it with alliance-check-toolkit before that though.)

**Comment by L-as** [comment]
Sent Jean-Paul Chaput an e-mail about perhaps upstreaming this work.

File: flake.nix [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+          ${elem} = pkgs.${"coriolis-${elem}"};
+        }) {} components
+        // {
+          test = pkgs.python2.buildEnv.override {
+            extraLibs = [ pkgs.coriolis-unicorn pkgs.coriolis-cumulus ];
+          };
+        }
+      );
+
+      defaultPackage = forAllSystems (system: self.packages.${system}.unicorn);
```
Comment (DieracDelta) [inline flake.nix:111]:
```suggestion
      # CORIOLIS_TOP must be set before execution
      # example: CORIOLIS_TOP="$(mktemp -d)" && ./result/bin/cgt
      defaultPackage = forAllSystems (system: self.packages.${system}.unicorn);
```

File: nix/alliance-check-toolkit.nix [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+  postPatch = "patchShebangs .";
+  configurePhase = "export CHECK_TOOLKIT=\"$PWD\"";
+  buildPhase = ''
+    cd benchs
+    ../bin/go.sh
+  '';
+  installPhase = "touch $out";
+  fixupPhase = "true";
+}
```
Comment (DieracDelta) [inline nix/alliance-check-toolkit.nix]: Is this meant as a check? As in "if it builds, the check passes"

Comment (L-as) [reply nix/alliance-check-toolkit.nix]: Yep. I'm going to make it a check in the flake too.

Comment (DieracDelta) [reply nix/alliance-check-toolkit.nix]: Yeah, that makes sense. The resulting `result` file isn't marked as executable for me (and by the looks of it is an empty file).

File: nix/tutorial.nix [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+    src = ../tutorial;
+
+    buildInputs = [
+      coriolis-vlsisapd coriolis-hurricane coriolis-crlcore
+      coriolis-lefdef libxml2 qt4
+    ];
+    nativeBuildInputs = [ doxygen ];
+    pythonImportsCheck = [ "Tutorial" ];
+  };
+in generic pkg
```
Comment (DieracDelta) [inline nix/tutorial.nix:19]: I'm getting a failure here: 
```
❯ ./tutorial
[ERROR] The <Cfg> python module or symbol cannot be loaded.
        Please check the integrity of the <coriolis> package.
[ERROR] The <No module named Cfg> shared library cannot be loaded.
        Under RHEL 6, you must be under devtoolset-2.
        (scl enable devtoolset-2 bash)
```
I'm guessing this is only used as a dependency of unicorn?

File: nix/unittests.nix [hunk 7]
```diff
@@ -0,0 +1,32 @@
+{ version, meta }:
```
Comment (DieracDelta) [inline nix/unittests.nix:1]: might be worth adding this package as a check as well

File: nix/tutorial.nix (hunk 6 above)
Comment (L-as) [reply nix/tutorial.nix:19]: Yeah, I'm not sure how you're supposed to use it actually.

File: nix/unittests.nix (hunk 7 above)
Comment (L-as) [reply nix/unittests.nix:1]: Will do.

**Comment by DieracDelta** [comment]
Well done!! LGTM with one caveat. Very clean stuff IMO. We should definitely try to upstream if possible. I think this could make a huge difference in the usability of coriolis.

I have one question: when running the tool, I can't seem to get the tutorials to run. Any ideas on how we could get this to work/why this doesn't run?

**Comment by L-as** [comment]
I'm not sure, perhaps 1.4 in https://libre-soc.org/HDL_workflow/coriolis2/ is how to use it? Perhaps it has to be used through `cgt`?

**Comment by DieracDelta** [comment]
To be a bit clearer on the steps involved:
```nix
nix build .
export CORIOLIS_TOP="$(mktemp -d)"
./result/bin/cgt
```

Then I select either of the two tutorials presented by the dropdown:
image
image

This drops the following warning:
```
[WARNING] Unicorn system configuration file:
  </tmp/tmp.Czu4VZvnCe/lib/python2.7/site-packages/unicorn/unicornInit.py> not found.
  ```
I think this is the step process they're describing in that readme you've linked to.

**Comment by DieracDelta** [comment]
It looks like `unicornInit.py` is in `result`. When I export `CORIOLIS_TOP` to be `result`, then cgt fails to start. I get the following errors:
```
❯ ./result/bin/cgt
  o  Locating configuration directory:
     - "/home/jrestivo/SoN/reviews/coriolis/result/etc/coriolis2"
     - ".../lib/python2.7/site-packages/symbolic/cmos/technology.py".
     - ".../lib/python2.7/site-packages/symbolic/cmos/display.py".
     - ".../lib/python2.7/site-packages/symbolic/cmos/analog.py".
     - ".../lib/python2.7/site-packages/symbolic/cmos/alliance.py".
     - ".../lib/python2.7/site-packages/symbolic/cmos/etesian.py".
     - ".../lib/python2.7/site-packages/symbolic/cmos/kite.py".
     - ".../lib/python2.7/site-packages/symbolic/cmos/plugins.py".
     - ".../lib/python2.7/site-packages/symbolic/cmos/stratus1.py".
```

 which makes sense. It looks like they're using relative paths. This might be entirely red-herring of what the actual reason is, though.

**Comment by L-as** [comment]
@DieracDelta From what I can tell, you're supposed to have some settings in the current directory. What happens if you copy it to your cwd?

**Comment by L-as** [comment]
Note to myself: I need to add the license attributes for the packages.

**Comment by DieracDelta** [comment]
@L-as I didn't have much luck trying that (same thing).

I don't think getting the tutorial to work is super high priority just as long as it can do its job for libre-SoC. The code quality is high and it seems to run fine. 

Did you have any luck contacting upstream?

**Comment by L-as** [comment]
I haven't contacted them since Jean-Paul Chaput told me he was taking a vacation for 2 weeks. I should probably contact him again.

**Comment by aaronjanse** [comment]
Would it be possible to do the `coriolis -> coriolis2` rename programmatically (e.g. via something like `sed`) instead of writing a large patch?

**Comment by L-as** [comment]
Why would you rename it?

**Comment by aaronjanse** [comment]
Oh, I was referring to the diff. For example:

https://github.com/ngi-nix/coriolis/pull/1/files#diff-0daba194b4ca6ffabf28a934928d4fbff89dda356c20809e86581c563e0d8fe9R21

**Comment by L-as** [comment]
Ah, I mean yeah, I could do a sed instead, but the problem is that this is actually an upstream bug, i.e. the CMake files are broken. When I eventually upstream this, I imagine they want the fixed CMake files too.

**Comment by aaronjanse** [comment]
Oh got it. Sounds good

**Comment by L-as** [comment]
I merged with upstream, not sure if it still works lol

**Comment by aaronjanse** [comment]
The `flake.lock` seems to [reference a commit that does not exist](https://github.com/ngi-nix/coriolis/blob/6c44a7a8a1293fa32d51bfad4b37d766eb771679/flake.lock#L36-L43), causing the following error:
```
error: unable to download 'https://api.github.com/repos/L-as/nixpkgs/tarball/11f9ff27bc66bf653ccbbfae1412e6f5705e4a5c': HTTP error 404 ('')

       response body:

       404: Not Found
(use '--show-trace' to show detailed location information)
```

**Comment by L-as** [comment]
Ah! I've deleted the branch and forgot to update the input here. My PRs have been merged now after all.

**Comment by L-as** [comment]
I'm in contact with upstream and trying to merge it upstream so I'll fix this.

**Comment by L-as** [comment]
I made a PR: https://gitlab.lip6.fr/vlsi-eda/coriolis/-/merge_requests/8

**Comment by L-as** [comment]
This has been merged! https://gitlab.lip6.fr/vlsi-eda/coriolis/-/merge_requests/8

**Pull request merged**